File: lesson34/lesson34.js
// JavaScript có gì thay thế?

// 1. public (mặc định) === Tất cả property trong JS mặc định là public
class Person {
  constructor(name) {
    this.name = name; // public
  }
}

const p = new Person("Eric");
console.log(p.name); //  truy cập được

// 2. private (ES2022) === Sử dụng ký hiệu # để định nghĩa private property
class Person {
  #ssn;

  constructor(ssn) {
    this.#ssn = ssn;
  }

  getSSN() {
    return this.#ssn;
  }
}

const p = new Person("123");
console.log(p.getSSN()); // truy cập được thông qua phương thức getter
// console.log(p.#ssn);  Syntax Error

// 3. protected (ES2022) === JS không có protected, nhưng có thể “giả lập” bằng cách sử dụng convention (quy ước) hoặc Symbol
class Person {
  _name; // convention thôi

  constructor(name) {
    this._name = name;
  }
}

class Student extends Person {
  getName() {
    return this._name; // vẫn dùng được
  }
}

// 4. static (ES2022) ===  static = thuộc về class, không thuộc về object
class Person {
  static species = "Homo sapiens";

  constructor(name) {
    this.name = name;
  }
}

console.log(Person.species); // truy cập được thông qua class
const p = new Person("Eric");
console.log(p.name); // truy cập được thông qua object

File: lesson34/lesson34.java
package lesson34;

// 1. public === Có thể truy cập từ bất kỳ đâu
public class lesson34 {
    public String name;

    public lesson34(String name) {
        this.name = name;
    }
}


// 2. private === Chỉ dùng trong chính class đó
class lesson34 {
    private String ssn;

    public String getSsn() {
        return ssn;
    }
}


// 3. protected === Có thể truy cập trong cùng package và các lớp con ở package khác
class lesson34 {
    protected String name;
}


class Student extends lesson34 {
    void show() {
        System.out.println(name); // OK
    }
}


// 4. Default (không ghi gì) === Chỉ dùng trong cùng package, không thể truy cập từ package khác
class lesson34 {
    String name; // default
}
